content/blog/2019-10-23-divergence-optimizations/index.md

```diff
@@ -0,0 +1,231 @@
+
+|     Test      | Description | Expected Optimization
+| ------------- | ----------- | ---------- 
+| vvadd | `vload` followed by a `vadd` | None, due to `vload` |
+| Unique scalars  | Unique values written to vector (`s2v`), then `vadd` | None, due to unique values |
+| Redundant scalars  | Redundant values written to vector, then `vadd` | All instructions should be scalar |
+| Partially divergent  | Both convergent and divergent instructions | Optimize convergent instructions and add `s2v` as needed before divergent instructions |
+| Predication - Unique   | Divergent predicated vector instructions | None, all divergent |
+| Predication - Redundant | Convergent predicated vector instructions | Optimize convergent instructions with divergent predication and remove predication when possible.
+
+### Performance
+
+#### Metric
+
+Our evaluation metric on the imaginary hardware is the number of ALU ops required by the program. Generally, each vector instruction consumes four ALU ops and each scalar instruction consumes a single ALU op. In this model, a scalar instruction is exactly four times as energy efficient as a redundant vector instruction. We argue that this is a good proxy metric for energy consumption if only the dynamic energy consumption of the ALU is considered and no other parts of the processor are considered (like memory access and on-chip network).
+
+#### Benchmarks
+
+We evaluate the effectiveness of the divergence optimizations on synthetic benchmarks. We take benchmark inspiration from the examples in [these](https://ieeexplore.ieee.org/document/6494995) [papers](https://hal.inria.fr/hal-00909072v2/document). All benchmarks have a 2D loop nest. We vectorize over each outer loop and unroll over each inner loop because we do not support most control flows. We manually unroll the inner loop twice for each benchmark as it allows us to get a sense of the dynamic ALU ops without actually running the program. The number of ALU ops for the baseline and optimized version of each benchmark is shown in the table below.
+
+|     Benchmark    | Description | Baseline Ops | Optimized Ops | Improvement (%) |
+| ------------- | ------------| ------------- | ------------ | ---------- |
+| FIR  | 2D FIR filter | 60 | 53 | 12
+| FIR-pred | 2D FIR filter with single conditional | 87 | 81 | 7
+| Synthetic | Sum of (a[outer] * b[inner]) / c[inner] | 53 | 39 | 26
+
+The optimization does lead to improvement in the number of ALU Ops for the listed benchmarks. It's hard to say exactly what a fair baseline would be because we don't know what a SPMD front-end would actually emit. For example, loads that aren't contiguous use `gather` in our baseline. In the case where each address is the same (i.e. indexed by the inner loop iterator), the `gather` can be turned into a `lw`. It's possible that this is obvious enough for a SPMD front-end to do automatically. We don't have a SPMD to Bril compiler, so we can't truly quantify a realistic improvement.
```

Comment:
```suggestion
The optimization does lead to improvement in the number of ALU Ops for the listed benchmarks. It's hard to say exactly what a fair baseline would be because we don't know what a SPMD front-end would actually emit. For example, loads that aren't contiguous use `gather` in our baseline. In the case where each address is the same (i.e., indexed by the inner loop iterator), the `gather` can be turned into a `lw`. It's possible that this is obvious enough for a SPMD front-end to do automatically. We don't have a SPMD to Bril compiler, so we can't truly quantify a realistic improvement.
```